apps/omg_child_chain/lib/child_chain.ex

```diff
@@ -24,23 +24,38 @@ defmodule OMG.ChildChain do
   alias OMG.Block
   alias OMG.ChildChain.FeeServer
   alias OMG.ChildChain.FreshBlocks
+  alias OMG.ChildChain.Transaction.Metrics
+  alias OMG.ChildChain.Transaction.Submitter
   alias OMG.Fees
   alias OMG.Fees.FeeFilter
-  alias OMG.State
   alias OMG.State.Transaction
 
-  @type submit_error() :: Transaction.Recovered.recover_tx_error() | State.exec_error() | :transaction_not_supported
-
-  @spec submit(transaction :: binary) ::
+  @spec submit(transaction :: binary, submitter :: Submitter) ::
           {:ok, %{txhash: Transaction.tx_hash(), blknum: pos_integer, txindex: non_neg_integer}}
-          | {:error, submit_error()}
-  def submit(transaction) do
-    with {:ok, recovered_tx} <- Transaction.Recovered.recover_from(transaction),
-         :ok <- is_supported(recovered_tx),
-         {:ok, fees} <- FeeServer.accepted_fees(),
-         fees = Fees.for_transaction(recovered_tx, fees),
-         {:ok, {tx_hash, blknum, tx_index}} <- State.exec(recovered_tx, fees) do
-      {:ok, %{txhash: tx_hash, blknum: blknum, txindex: tx_index}}
+          | {:error, Submitter.submit_error_t()}
+  def submit(transaction, submitter \\ Submitter, metrics \\ Metrics) do
```

Comment:
have you considered moving one level up on the webserver? you could do this without interfering/changing the flow on the webserver level, no?

Reply:
I did think about this a lot. I'd prefer to keep the metric where it is for now. 

Happy to talk it out a bit more since this will set a precedent for many metrics follow.  But I don't want to drag the conversation on too long so let's decide...today please. 😄 Doesn't have to be perfect, it can be changed later.

@unnawut please feel free to break the tie.

My reasoning is below just for context.

It might make sense to have in both places. It depends. The entire 'trace' might have 4 places for the metric: WatcherRPC -> Watcher -> ChildChainRPC -> ChildChain. Let's wait and see what we want from it.

But I believe given the info we have at hand, having the metric as close to the code as possible, is the best guarantee we have at capturing the metric most accurately. It future proofs it, against new code calling `submit()`. It makes no assumptions about how or where `ChildChain.submit()` is called.

Putting it at the RPC only, makes a lot of assumptions about how and where `ChildChain.submit()` is called from, eg: it will always and only ever be called from ``ChildChainRPC.submit()`.

This did require a minor code refactoring though. The refactoring is highly beneficial for testing purposes (not just metrics testing), so I'm comfortable to make it. See the test file for the level of granularity the both the metric and metadata can be captured. RPC layer can't cover it as accurately.

What are there benefits of putting the metric at the RPC layer?

Happy to discuss, but how about let's make the deadline today and just move on until we know more.